Fix all build warnings when compiling with clang

vickio opened the pull request:
I can't test with any other build systems, but for macOS 10.12 compiling with clang, this fixes all the build warnings. I didn't correct the warnings for the tests, just the main build.

vickio commented on contrib/epee/include/md5_l.h:
```diff
@@ -74,7 +74,7 @@ namespace md5
 	static void MD5Update( MD5_CTX *context, const unsigned char *input, unsigned int inputLen );
 	static void MD5Final ( unsigned char digest[16], MD5_CTX *context );
-	static void hmac_md5(const unsigned char* text, int text_len, const unsigned char* key, int key_len, unsigned char *digest);
+	static inline void hmac_md5(const unsigned char* text, int text_len, const unsigned char* key, int key_len, unsigned char *digest);
```
This function is unused.

vickio commented on CMakeLists.txt:
```diff
@@ -350,7 +350,7 @@ else()
   endif()
   if(CMAKE_C_COMPILER_ID STREQUAL "Clang")
-    set(WARNINGS "${WARNINGS} -Wno-deprecated-register -Wno-error=mismatched-tags -Wno-error=null-conversion -Wno-overloaded-shift-op-parentheses -Wno-error=shift-count-overflow -Wno-error=tautological-constant-out-of-range-compare -Wno-error=unused-private-field -Wno-error=unneeded-internal-declaration")
+    set(WARNINGS "${WARNINGS} -Wno-deprecated-register -Wno-overloaded-shift-op-parentheses -Wno-missing-braces")
```
`missing-braces` was triggering lots of warnings from `rctTypes.h` but the code is ok.

NanoAkron commented:
I'm not sure that hiding build warnings is the right way to go. Why not fix the things that are causing the warnings? It's like 'fixing' your car by disabling the red engine warning light. 

Others may have different philosophies.

vickio commented:
@NanoAkron I agree, except that these warnings seem to be a clang bug: https://llvm.org/bugs/show_bug.cgi?id=21629. I am looking to into a workaround now though that might allow `missing-braces` to stay.

NanoAkron commented:
Ah I see :)

moneromooo-monero commented:
I'd rather stuff be splt in logically separate things when it makes sense (ie, the git stuff has nothing to do with the class/struct changes). A few comments needing attention.

vickio commented:
@moneromooo-monero ok, do you mean separate commits?

moneromooo-monero commented:
Yes. Put things together that form a whole.

moneromooo-monero commented:
I'm particularly interested in your change to the git command. I think it's wrong, but I want to know your reason for changing it.

vickio commented:
Yeah, turns out my git repo was messed up and `git rev-list --tags` was failing, but a new clone fixed it. I removed the git change and split the other changes into separate commits for each warning type that the changes address.

There are still some linking errors that have to do with a boost visibility setting mismatch, but it seems to depend on how your boost library was built. Maybe there needs to be an option that can be set to match the boost visibility? Or some way to detect it? I really don't know much about boost.

moneromooo-monero commented:
What do you mean about boost visibility ? Symbol visibility (ie, public, private, internal),or other ?

vickio commented on src/wallet/wallet2.cpp:
```diff
@@ -5035,27 +5035,28 @@ std::string wallet2::make_uri(const std::string &address, const std::string &pay
   {
-    uri += (n_fields++ ? "&" : "?") + std::string("tx_payment_id=") + payment_id;
+    uri += "&" + std::string("tx_payment_id=") + payment_id;
+    n_fields++;
```
I fixed a logic error in the last commit, maybe it's ok now?

moneromooo-monero replied:
No. The first separator between a URL's path and the first field must be ?.

vickio commented on external/db_drivers/liblmdb/mdb.c:
```diff
@@ -10707,9 +10707,11 @@ mdb_reader_list(MDB_env *env, MDB_msg_func *func, void *ctx)
+			}
 			if (first) {
 				first = 0;
 				rc = func("    pid     thread     txnid\n", ctx);
```
Right, I wondered how you were handling the external libs. I'll remove it from this file it there.

vickio replied:
Looks like this one will not be fixed upstream:
http://www.openldap.org/its/index.cgi/Incoming?id=8407
> The code is correct. Closing this ITS.
> 
> It is not the compiler's job to scold programmers about style.

vickio commented:
About boost, I mean the linker throws a ton of these types of warnings (you can see them in the buildbot OS X compiler logs):

> ld: warning: direct access in function 'boost::archive::detail::archive_serializer_map<boost::archive::binary_iarchive>::insert(boost::archive::detail::basic_serializer const*)' from file '/usr/local/lib/libboost_serialization.a(binary_iarchive.o)' to global weak symbol 'boost::serialization::singleton<boost::archive::detail::extra_detail::map<boost::archive::binary_iarchive> >::get_instance()::t' from file 'CMakeFiles/daemon.dir/daemon.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.

I researched it little but I don't really understand what's going on.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -5035,27 +5035,28 @@ std::string wallet2::make_uri(const std::string &address, const std::string &pay
   {
     // URI encoded amount is in decimal units, not atomic units
-    uri += (n_fields++ ? "&" : "?") + std::string("tx_amount=") + cryptonote::print_money(amount);
+    uri += (n_fields++ == 0 ? "&" : "?") + std::string("tx_amount=") + cryptonote::print_money(amount);
```
It was right originally.

vickio replied:
Ah, of course. It was right originally except for `tx_payment_id`.

moneromooo-monero commented:
OK, though I'm not sure why you persist in having a custom bit of code payment id, as it makes the code vulnerable to reordering fields around, and makes the test different from the others. But it's fine I guess.

vickio commented:
I guess it seemed weird having a condition on the first parameter. I just made it consistent with the others.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -5035,27 +5035,27 @@ std::string wallet2::make_uri(const std::string &address, const std::string &pay
   if (!payment_id.empty())
   {
-    uri += (n_fields++ ? "&" : "?") + std::string("tx_payment_id=") + payment_id;
+    uri += (n_fields++ > 0 ? "&" : "?") + std::string("tx_payment_id=") + payment_id;
```
Those changes are a bit pointless, no ? They just halve the range in which the code will work :) Not that the range change matters in practice, though. Might as well remove them, and maybe make n_fields unsigned while at it.

vickio replied:
With the change from `bool` to `int`, the explicit condition was added for clarity, since the question is whether or not it's the first parameter. I removed the comparison and made the counter unsigned.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -5035,7 +5035,7 @@ std::string wallet2::make_uri(const std::string &address, const std::string &pay
 
   std::string uri = "monero:" + address;
-  bool n_fields = 0;
+  uint8_t n_fields = 0;
```
I hate to be so picky, but 256 parameters is plausibly reached, as this will be automated. Please use, eg, unsigned int.

vickio replied:
Ok. I don't understand what you mean, but I did change it.

NanoAkron commented on CMakeLists.txt:
```diff
@@ -350,7 +350,6 @@ else()
     set(WARNINGS_AS_ERRORS_FLAG "-Werror")
   endif()
   if(CMAKE_C_COMPILER_ID STREQUAL "Clang")
```
Do you need this conditional if() if you're removing what it actually does?

vickio replied:
Yeah, `if(ARM)` is still contained in it.

NanoAkron replied:
Oh yeah

moneromooo-monero commented:
ok
